tests/Functional/ResultCacheTest.php

```diff
@@ -225,6 +225,64 @@ public function testFetchAllSavesCache(): void
         self::assertCount(1, $layerCache->fetch('testcachekey'));
     }
 
+    public function testCacheQueriedOnlyOnceForCacheMiss(): void
+    {
+        $layerCache = $this->createMock(ArrayCache::class);
+        $layerCache->expects(self::once())
```

Comment:
It can't be a unit test as it covers the full use case. Right now, there are two invocations of `fetch` - in different places of the library. This test checks, that `fetch` from real cache is executed only once. 

In my PR it is executed in the `Connection::executeCacheQuery` method. However, it might be executed anywhere and moved somewhere else in the future.

Reply:
I may change it to `ArrayCache`, but:
1) it will depend on the `ArrayCache` implementation, which is a part of another project - `Doctrine/Common`.
2) essentially `ArrayCache` is a mock by itself. I can hardly imagine why this class might be used in the production environment.
3) `dbal` must accept any object, that implements `Doctrine\Common\Cache\Cache` interface and work with it. This class is external relative to `dbal`.

Functional testing performs tests of the whole library by sending input to its end-user endpoint and examining output (without checking internal structure). In our case, cache-object is just another type of the input. And we're checking whether `dbal` uses it correctly.

Reply:
What's your take on "Don't mock what you don't own?"
To me, as you pointed out in 2. `ArrayCache` is already a mock, but with a few differences:
a) it's easier to use, you don't have to write any expectations, only `new ArrayCache()`
b) it's closer to reality although it probably won't be used in production (or won't it? you don't know that, could be useful in a long-running process)
c) it is not controlled by you, and a change in it may break your tests; and if that's the case _you want to know_ you don't want to wait until production to notice it breaks.

Reply:
I don't get this question and how it is related to this topic.

Let me explain one more time:

`dbal` has the "contract": it will accept any object that implements `Doctrine\Common\Cache\Cache` interface and will use it independenty of it's internal structure to store and retrieve cache entries. In most unit and functional tests `Doctrine\Common\Cache\ArrayCache` is used. But it's done just for simplicity. `dbal` must be able to work with any cache provider. 

I'm mocking my own object, that implements `Doctrine\Common\Cache\Cache` interface, because in this particular case it's easier. 

Why `Doctrine\Common\Cache\ArrayCache` is harder to use? The only way to get the number of fetches from it is to retrieve statistics by the `getStats()` method. But consider running this code:

```php
$cache = new ArrayCache();
$cache->fetch('test_key');
self::assertEquals(1, $cache->getStats()[Cache::STATS_MISSES]);
```

It will fail. The right value for `Cache::STATS_MISSES` after one `fetch()` is 2. It's because of their internal structure (the first cache miss is due to fetching namespace version - but it doesn't matter here). So instead of my current test code 

```php
$layerCache->expects(self::once())->method('fetch')
```

I will have to write the following code 

```php
self::assertEquals(2, $layerCache->getStats()[Cache::STATS_MISSES]);
```

It's completely unclear why I'm using 2 here, while the whole idea of this patch is to remove the double fetch. And if they remove this "namespace version" feature in the `ArrayCache` (as it useless in this case), this test will fail. 

TL;DR:
1) This test is functional as it treats `dbal` as a whole and doesn't depend on its internal structure
2) Mock is used because it's easier to calculate number of real fetches. While `ArrayCache` is quite simple, its statistics is distorted.

Reply:
Oh I see, you really need a mock since you are adding expectations to it.

> essentially ArrayCache is a mock by itself.

It's more like a stub in fact, that's what got me confused. It's easier to use as a stub, but can't be used as a mock since you can't know which methods were called,  and instead just get some distorted stats.

Reply:
Why does interaction between the units of code (the cache and the underlying statement) need to be tested in a functional way? It's obviously a unit-level requirement, not a functional requirement.

Reply:
> Why does interaction between the units of code

Any interaction is an "interaction between units of code". The question is where these units of code reside. Cache is not part of DBAL.


> A cache is a non-functional unit by definition since it's not meant to change the behavior of the system. 

Cache DOES change the behavior of the system. The system is working differently depending on what the cache object returns. 

I see you don't understand (or don't want to understand) why this test is functional. I can delete it if you want. Or I can copy it to the "tests/Cache" directory.

Reply:
> I see you don't understand (or don't want to understand)

Please treat others more kindly, especially when they take that much time to answer to you. Assuming either ignorance or malice like this will not help.

As for your definition of behavior, it's debatable. To me, it would be more about direct inputs and outputs / the public interface of the system. Whether the system calls systems external systems to create the output should not be relevant in a functional test IMO.